x-pack/plugins/infra/server/usage/usage_collector.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { InfraNodeType } from '../../common/graphql/types';
+import { KbnServer } from '../kibana.index';
+
+const KIBANA_REPORTING_TYPE = 'infraops';
+
+interface InfraopsSum {
+  infraopsHosts: number;
+  infraopsDocker: number;
+  infraopsKubernetes: number;
+  logs: number;
+}
+
+export class UsageCollector {
+  public static getUsageCollector(server: KbnServer) {
+    const { collectorSet } = server.usage;
+
+    return collectorSet.makeUsageCollector({
+      type: KIBANA_REPORTING_TYPE,
+      fetch: async () => {
+        return this.getReport();
+      },
+    });
+  }
+
+  public static countNode(nodeType: InfraNodeType) {
+    const bucket = this.getBucket();
+    this.maybeInitializeBucket(bucket);
+
+    switch (nodeType) {
+      case InfraNodeType.pod:
+        this.counters[bucket].infraopsKubernetes += 1;
+        break;
+      case InfraNodeType.container:
+        this.counters[bucket].infraopsDocker += 1;
+        break;
+      default:
+        this.counters[bucket].infraopsHosts += 1;
+    }
+  }
+
+  public static countLogs() {
+    const bucket = this.getBucket();
+    this.maybeInitializeBucket(bucket);
+    this.counters[bucket].logs += 1;
+  }
+
+  private static counters: any = {};
+  private static BUCKET_SIZE = 3600; // seconds in an hour
+  private static BUCKET_NUMBER = 24; // report the last 24 hours
+
+  private static getBucket() {
+    const now = Math.floor(Date.now() / 1000);
+    return now - (now % this.BUCKET_SIZE);
+  }
+
+  private static maybeInitializeBucket(bucket: any) {
+    if (!this.counters[bucket]) {
+      this.counters[bucket] = {
+        infraopsHosts: 0,
+        infraopsDocker: 0,
+        infraopsKubernetes: 0,
+        logs: 0,
+      };
+    }
+  }
+
+  private static getReport() {
+    const keys = Object.keys(this.counters);
+    // only keep the newest BUCKET_NUMBER buckets
+    keys.forEach(key => {
+      if (parseInt(key, 10) < this.getBucket() - this.BUCKET_SIZE * (this.BUCKET_NUMBER - 1)) {
```

Comment:
Calling `this.getBucket()` separately means that, if `getReport()` is called exactly on a bucket boundary, we might end up with an inconsistent length of the history between different counters.

The risk of hitting such a situation might not be too high. At the same time it would not be surprising for stats collection scripts to run on the full hour in naive cron configurations.

In any case, moving this static calculation outside of the loop should prevent that inconsistency.

Reply:
:+1: Good point, changed.